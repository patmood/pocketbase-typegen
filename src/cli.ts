import dotenv from "dotenv"

import type { CollectionRecord, Options } from "./types"
import { fromDatabase, fromJSON, fromURL } from "./schema"


import { generate } from "./lib"
import { saveFile } from "./utils"

export async function main(options: Options) {
  let schema: Array<CollectionRecord>
  if (options.db) {
    schema = await fromDatabase(options.db)
  } else if (options.json) {
    schema = await fromJSON(options.json)
  } else if (options.url) {
    schema = await fromURL(options.url, options.email, options.password)
  } else if (options.env) {
    const path: string = typeof options.env === "string"
      ? options.env
      : ".env"
    dotenv.config({ path: path })
    if (!process.env.PB_TYPEGEN_URL || !process.env.PB_TYPEGEN_EMAIL || !process.env.PB_TYPEGEN_PASSWORD) {
      return console.error(
        "Missing environment variables. Check options: pocketbase-typegen --help"
      )
    }
    schema = await fromURL(
      process.env.PB_TYPEGEN_URL,
      process.env.PB_TYPEGEN_EMAIL,
      process.env.PB_TYPEGEN_PASSWORD
    )
  } else {
    return console.error(
      "Missing schema path. Check options: pocketbase-typegen --help"
    )
  }
  const typeString = generate(schema, {
    sdk: options.sdk ?? true
  })
  await saveFile(options.out, typeString)
  return typeString
}
