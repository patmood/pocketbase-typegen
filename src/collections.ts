import { toPascalCase } from "./utils"

export function createCollectionEnum(collectionNames: Array<string>): string {
  const collections = collectionNames
    .map((name) => `\t${toPascalCase(name)} = "${name}",`)
    .join("\n")
  const typeString = `export enum Collections {
${collections}
}`
  return typeString
}

export function createCollectionRecords(
  collectionNames: Array<string>
): string {
  const nameRecordMap = collectionNames
    .map((name) => `\t${name}: ${toPascalCase(name)}Record`)
    .join("\n")
  return `export type CollectionRecords = {
${nameRecordMap}
}`
}

export function createCollectionResponses(
  collectionNames: Array<string>
): string {
  const nameRecordMap = collectionNames
    .map((name) => `\t${name}: ${toPascalCase(name)}Response`)
    .join("\n")
  return `export type CollectionResponses = {
${nameRecordMap}
}`
}

export function createTypedPocketbase(collectionNames: Array<string>): string {
  const nameRecordMap = collectionNames
    .map(
      (name) =>
        `\tcollection(idOrName: '${name}'): RecordService<${toPascalCase(
          name
        )}Response>`
    )
    .join("\n")
  return `export type TypedPocketBase = PocketBase & {
${nameRecordMap}
}`
}
