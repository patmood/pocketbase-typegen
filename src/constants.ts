export const EXPORT_COMMENT = `/**
* This file was @generated using pocketbase-typegen
*/`
export const RECORD_TYPE_COMMENT = `// Record types for each collection`
export const RESPONSE_TYPE_COMMENT = `// Response types include system fields and match responses from the PocketBase API`
export const EXPAND_GENERIC_NAME = "expand"
export const DATE_STRING_TYPE_NAME = `IsoDateString`
export const RECORD_ID_STRING_NAME = `RecordIdString`
export const HTML_STRING_NAME = `HTMLString`
export const ALIAS_TYPE_DEFINITIONS = `// Alias types for improved usability
export type ${DATE_STRING_TYPE_NAME} = string
export type ${RECORD_ID_STRING_NAME} = string
export type ${HTML_STRING_NAME} = string`

export const BASE_SYSTEM_FIELDS_DEFINITION = `// System fields
export type BaseSystemFields<T = never> = {
\tid: ${RECORD_ID_STRING_NAME}
\tcreated: ${DATE_STRING_TYPE_NAME}
\tupdated: ${DATE_STRING_TYPE_NAME}
\tcollectionId: string
\tcollectionName: Collections
\texpand?: T
}`

export const AUTH_SYSTEM_FIELDS_DEFINITION = `export type AuthSystemFields<T = never> = {
\temail: string
\temailVisibility: boolean
\tusername: string
\tverified: boolean
} & BaseSystemFields<T>`

export const VIEW_SYSTEM_FIELDS_DEFINITION = `export type ViewSystemFields<T = never> = {
  \tid: ${RECORD_ID_STRING_NAME}
  \tcollectionId: string
  \tcollectionName: Collections
  \texpand?: T
}`
