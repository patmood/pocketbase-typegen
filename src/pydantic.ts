import { CollectionRecord, FieldSchema, pydanticTypeMap } from "./types"
import { toPascalCase } from "./utils"

export function generatePydantic(schema: Array<CollectionRecord>): string {
  const imports = `from pydantic import BaseModel, Field, ConfigDict
from typing import Optional, List, Dict, Any, Literal, TypeVar, Generic
from datetime import datetime
from enum import Enum, auto`

  const models = schema
    .map((collection) => {
      // Generate enums for select fields first
      const enums = collection.fields
        .filter(
          (field) =>
            field.type === "select" && field.values && field.values.length > 0
        )
        .map((field) => generateEnum(collection.name, field))
        .join("\n\n")

      // Generate the actual model
      const fields = collection.fields
        .map((field) => generatePydanticField(collection.name, field))
        .join("\n    ")

      const baseClass =
        collection.type === "auth" ? "AuthSystemFields" : "BaseSystemFields"

      return `
${enums}

class ${toPascalCase(collection.name)}Base(${baseClass}):
    """Base model for ${collection.name} collection"""
    ${fields}

class ${toPascalCase(collection.name)}Create(${toPascalCase(
        collection.name
      )}Base):
    """Create model for ${
      collection.name
    } collection (used for creation/updates)"""
    pass

class ${toPascalCase(collection.name)}(${toPascalCase(collection.name)}Base):
    """Full model for ${collection.name} collection including system fields"""
    id: str
    created: datetime
    updated: datetime
    
    model_config = ConfigDict(from_attributes=True)`
    })
    .join("\n\n")

  // Add base system field classes
  const systemFields = `
class BaseSystemFields(BaseModel):
    """Base system fields included in all collections"""
    id: str
    
class AuthSystemFields(BaseSystemFields):
    """Additional system fields for auth collections"""
    email: str
    email_visibility: bool
    username: str
    verified: bool`

  // Add collection mapping
  const collectionMapping = `
# Type mapping for all collections
Collections = {
    ${schema
      .map(
        (collection) => `"${collection.name}": ${toPascalCase(collection.name)}`
      )
      .join(",\n    ")}
}

# Type alias for any collection model
AnyCollection = ${schema
    .map((collection) => toPascalCase(collection.name))
    .join(" | ")}
`

  return `${imports}\n\n${systemFields}\n\n${models}\n\n${collectionMapping}\n`
}

function generateEnum(collectionName: string, field: FieldSchema): string {
  if (!field.values) return ""
  const enumName = `${toPascalCase(collectionName)}${toPascalCase(
    field.name
  )}Options`
  const values = field.values.map((v) => `    ${v} = "${v}"`).join("\n")

  return `class ${enumName}(str, Enum):
    """Options for ${field.name} field in ${collectionName} collection"""
${values}`
}

function generatePydanticField(
  collectionName: string,
  field: FieldSchema
): string {
  const typeStr = getPydanticType(collectionName, field)
  const optional = !field.required
  const fieldType = optional ? `Optional[${typeStr}]` : typeStr
  const defaultValue = optional ? " = None" : ""
  const description =
    field.type === "relation"
      ? `\n    """Reference to ${field.options?.collectionId} collection"""`
      : ""

  return `${field.name}: ${fieldType}${defaultValue}${description}`
}

function getPydanticType(collectionName: string, field: FieldSchema): string {
  if (field.type === "select" && field.values && field.values.length > 0) {
    const enumName = `${toPascalCase(collectionName)}${toPascalCase(
      field.name
    )}Options`
    if (field.maxSelect && field.maxSelect > 1) {
      return `List[${enumName}]`
    }
    return enumName
  }

  const baseType =
    pydanticTypeMap[field.type as keyof typeof pydanticTypeMap] || "Any"

  if (field.type === "json") {
    return "Dict[str, Any]"
  }

  if (field.maxSelect && field.maxSelect > 1) {
    return `List[${baseType}]`
  }

  return baseType
}
