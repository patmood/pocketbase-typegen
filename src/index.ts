#!/usr/bin/env node

import type { Options } from "./types"
import { main } from "./cli"
import { program } from "commander"
import { version } from "../package.json"

program
  .name("Pocketbase Typegen")
  .version(version)
  .description(
    "CLI to create typescript typings for your pocketbase.io records"
  )
  .option("-d, --db <char>", "path to the pocketbase SQLite database")
  .option(
    "-j, --json <char>",
    "path to JSON schema exported from pocketbase admin UI"
  )
  .option(
    "-u, --url <char>",
    "URL to your hosted pocketbase instance. When using this options you must also provide email and password options."
  )
  .option(
    "-e, --email <char>",
    "email for an admin pocketbase user. Use this with the --url option"
  )
  .option(
    "-p, --password <char>",
    "password for an admin pocketbase user. Use this with the --url option"
  )
  .option(
    "-o, --out <char>",
    "path to save the typescript output file",
    "pocketbase-types.ts"
  )
  .option(
    "--no-sdk",
    "remove the pocketbase package dependency. A typed version of the SDK will not be generated."
  )
  .option(
    "-e, --env [path]",
    "flag to use environment variables for configuration. Add PB_TYPEGEN_URL, PB_TYPEGEN_EMAIL, PB_TYPEGEN_PASSWORD to your .env file. Optionally provide a path to your .env file"
  )

program.parse(process.argv)
const options = program.opts<Options>()
main(options)
