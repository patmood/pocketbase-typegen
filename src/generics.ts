import { EXPAND_GENERIC_NAME } from "./constants"
import { FieldSchema } from "./types"

export function fieldNameToGeneric(name: string) {
  return `T${name}`
}

export function getGenericArgList(schema: FieldSchema[]): string[] {
  const jsonFields = schema
    .filter((field) => field.type === "json")
    .map((field) => fieldNameToGeneric(field.name))
    .sort()
  return jsonFields
}

export function getGenericArgStringForRecord(schema: FieldSchema[]): string {
  const argList = getGenericArgList(schema)
  if (argList.length === 0) return ""
  return `<${argList.map((name) => `${name}`).join(", ")}>`
}

export function getGenericArgStringWithDefault(
  schema: FieldSchema[],
  opts: { includeExpand: boolean }
): string {
  const argList = getGenericArgList(schema)

  if (opts.includeExpand && canExpand(schema)) {
    argList.push(fieldNameToGeneric(EXPAND_GENERIC_NAME))
  }

  if (argList.length === 0) return ""
  return `<${argList.map((name) => `${name} = unknown`).join(", ")}>`
}

// Does the collection have relation fields that can be expanded
export function canExpand(schema: FieldSchema[]) {
  return !!schema.find((field) => field.type === "relation")
}
