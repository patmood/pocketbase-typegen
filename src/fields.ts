import {
  DATE_STRING_TYPE_NAME,
  HTML_STRING_NAME,
  RECORD_ID_STRING_NAME,
} from "./constants"
import { getOptionEnumName, getOptionValues, sanitizeFieldName } from "./utils"

import { FieldSchema } from "./types"
import { fieldNameToGeneric } from "./generics"

/**
 * Convert the pocketbase field type to the equivalent typescript type
 */
export const pbSchemaTypescriptMap = {
  // Basic fields
  bool: "boolean",
  date: DATE_STRING_TYPE_NAME,
  autodate: DATE_STRING_TYPE_NAME,
  editor: HTML_STRING_NAME,
  email: "string",
  text: "string",
  url: "string",
  password: "string",
  number: "number",

  // Dependent on schema
  file: (fieldSchema: FieldSchema) =>
    fieldSchema.maxSelect && fieldSchema.maxSelect > 1 ? "string[]" : "string",
  json: (fieldSchema: FieldSchema) =>
    `null | ${fieldNameToGeneric(fieldSchema.name)}`,
  relation: (fieldSchema: FieldSchema) =>
    fieldSchema.maxSelect && fieldSchema.maxSelect === 1
      ? RECORD_ID_STRING_NAME
      : `${RECORD_ID_STRING_NAME}[]`,
  select: (fieldSchema: FieldSchema, collectionName: string) => {
    // pocketbase v0.8+ values are required
    const valueType = fieldSchema.values
      ? getOptionEnumName(collectionName, fieldSchema.name)
      : "string"
    return fieldSchema.maxSelect && fieldSchema.maxSelect > 1
      ? `${valueType}[]`
      : valueType
  },

  // DEPRECATED: PocketBase v0.8 does not have a dedicated user relation
  user: (fieldSchema: FieldSchema) =>
    fieldSchema.maxSelect && fieldSchema.maxSelect > 1
      ? `${RECORD_ID_STRING_NAME}[]`
      : RECORD_ID_STRING_NAME,
}

export function createTypeField(
  collectionName: string,
  fieldSchema: FieldSchema
): string {
  let typeStringOrFunc:
    | string
    | ((fieldSchema: FieldSchema, collectionName: string) => string)

  if (!(fieldSchema.type in pbSchemaTypescriptMap)) {
    console.log(`WARNING: unknown type "${fieldSchema.type}" found in schema`)
    typeStringOrFunc = "unknown"
  } else {
    typeStringOrFunc =
      pbSchemaTypescriptMap[
        fieldSchema.type as keyof typeof pbSchemaTypescriptMap
      ]
  }

  const typeString =
    typeof typeStringOrFunc === "function"
      ? typeStringOrFunc(fieldSchema, collectionName)
      : typeStringOrFunc

  const fieldName = sanitizeFieldName(fieldSchema.name)
  const required = fieldSchema.required ? "" : "?"

  return `\t${fieldName}${required}: ${typeString}`
}

export function createSelectOptions(
  recordName: string,
  fields: Array<FieldSchema>
): string {
  const selectFields = fields.filter((field) => field.type === "select")
  const typestring = selectFields
    .map(
      (field) => `export enum ${getOptionEnumName(recordName, field.name)} {
${getOptionValues(field)
  .map((val) => `\t"${getSelectOptionEnumName(val)}" = "${val}",`)
  .join("\n")}
}\n`
    )
    .join("\n")
  return typestring
}

export function getSelectOptionEnumName(val: string) {
  if (!isNaN(Number(val))) {
    // The value is a number, which cannot be used as an enum name
    return `E${val}`
  } else {
    return val
  }
}
