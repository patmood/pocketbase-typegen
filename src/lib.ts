import {
  ALIAS_TYPE_DEFINITIONS,
  AUTH_SYSTEM_FIELDS_DEFINITION,
  BASE_SYSTEM_FIELDS_DEFINITION,
  EXPAND_GENERIC_NAME,
  EXPORT_COMMENT,
  RECORD_TYPE_COMMENT,
  RESPONSE_TYPE_COMMENT,
  VIEW_SYSTEM_FIELDS_DEFINITION,
} from "./constants"
import { CollectionRecord, FieldSchema } from "./types"
import {
  canExpand,
  getGenericArgStringForRecord,
  getGenericArgStringWithDefault,
} from "./generics"
import { createSelectOptions, createTypeField } from "./fields"
import { getSystemFields, toPascalCase } from "./utils"

export function generate(results: Array<CollectionRecord>): string {
  const collectionNames: Array<string> = []
  const recordTypes: Array<string> = []
  const responseTypes: Array<string> = [RESPONSE_TYPE_COMMENT]

  results
    .sort((a, b) => (a.name <= b.name ? -1 : 1))
    .forEach((row) => {
      if (row.name) collectionNames.push(row.name)
      if (row.schema) {
        recordTypes.push(createRecordType(row.name, row.schema))
        responseTypes.push(createResponseType(row))
      }
    })
  const sortedCollectionNames = collectionNames

  const fileParts = [
    EXPORT_COMMENT,
    createCollectionEnum(sortedCollectionNames),
    ALIAS_TYPE_DEFINITIONS,
    BASE_SYSTEM_FIELDS_DEFINITION,
    AUTH_SYSTEM_FIELDS_DEFINITION,
    VIEW_SYSTEM_FIELDS_DEFINITION,
    RECORD_TYPE_COMMENT,
    ...recordTypes,
    responseTypes.join("\n"),
    createCollectionRecords(sortedCollectionNames),
  ]

  return fileParts.join("\n\n")
}

export function createCollectionEnum(collectionNames: Array<string>): string {
  const collections = collectionNames
    .map((name) => `\t${toPascalCase(name)} = "${name}",`)
    .join("\n")
  const typeString = `export enum Collections {
${collections}
}`
  return typeString
}

export function createCollectionRecords(
  collectionNames: Array<string>
): string {
  const nameRecordMap = collectionNames
    .map((name) => `\t${name}: ${toPascalCase(name)}Record`)
    .join("\n")
  return `export type CollectionRecords = {
${nameRecordMap}
}`
}

export function createRecordType(
  name: string,
  schema: Array<FieldSchema>
): string {
  const selectOptionEnums = createSelectOptions(name, schema)
  const typeName = toPascalCase(name)
  const genericArgs = getGenericArgStringWithDefault(schema, {
    includeExpand: false,
  })
  const fields = schema
    .map((fieldSchema: FieldSchema) => createTypeField(name, fieldSchema))
    .join("\n")

  return `${selectOptionEnums}export type ${typeName}Record${genericArgs} = {
${fields}
}`
}

export function createResponseType(
  collectionSchemaEntry: CollectionRecord
): string {
  const { name, schema, type } = collectionSchemaEntry
  const pascaleName = toPascalCase(name)
  const genericArgsWithDefaults = getGenericArgStringWithDefault(schema, {
    includeExpand: true,
  })
  const genericArgsForRecord = getGenericArgStringForRecord(schema)
  const systemFields = getSystemFields(type)
  const expandArgString = canExpand(schema) ? `<T${EXPAND_GENERIC_NAME}>` : ""

  return `export type ${pascaleName}Response${genericArgsWithDefaults} = ${pascaleName}Record${genericArgsForRecord} & ${systemFields}${expandArgString}`
}
