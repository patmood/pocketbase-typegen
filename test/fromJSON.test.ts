import { promises as fs } from "fs"
import { main } from "../src/cli"
import path from "path"

it("creates a type file from json schema", async () => {
  const out = path.resolve(__dirname, "pocketbase-types-example.ts")
  const result = await main({
    out,
    json: path.resolve(__dirname, "pb_schema.json"),
  })

  const fileOutput = await fs.readFile(out, { encoding: "utf-8" })
  expect(fileOutput).toEqual(result)
  expect(fileOutput).toMatchSnapshot()
})
