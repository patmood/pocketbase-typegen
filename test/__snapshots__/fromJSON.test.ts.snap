// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates a type file from json schema 1`] = `
"/**
* This file was @generated using pocketbase-typegen
*/

export enum Collections {
	Base = "base",
	CustomAuth = "custom_auth",
	Everything = "everything",
	MyView = "my_view",
	Posts = "posts",
	Users = "users",
}

// Alias types for improved usability
export type IsoDateString = string
export type RecordIdString = string
export type HTMLString = string

// System fields
export type BaseSystemFields<T = never> = {
	id: RecordIdString
	created: IsoDateString
	updated: IsoDateString
	collectionId: string
	collectionName: Collections
	expand?: T
}

export type AuthSystemFields<T = never> = {
	email: string
	emailVisibility: boolean
	username: string
	verified: boolean
} & BaseSystemFields<T>

export type ViewSystemFields<T = never> = {
  	id: RecordIdString
  	collectionId: string
  	collectionName: Collections
  	expand?: T
}

// Record types for each collection

export type BaseRecord = {
	field?: string
}

export type CustomAuthRecord = {
	custom_field?: string
}

export enum EverythingSelectFieldOptions {
	"optionA" = "optionA",
	"OptionA" = "OptionA",
	"optionB" = "optionB",
	"optionC" = "optionC",
	"option with space" = "option with space",
	"sy?mb@!$" = "sy?mb@!$",
}
export type EverythingRecord<Tanother_json_field = unknown, Tjson_field = unknown> = {
	text_field?: string
	number_field?: number
	bool_field?: boolean
	email_field?: string
	url_field?: string
	date_field?: IsoDateString
	select_field?: EverythingSelectFieldOptions
	json_field?: null | Tjson_field
	another_json_field?: null | Tanother_json_field
	file_field?: string
	three_files_field?: string[]
	user_relation_field?: RecordIdString
	custom_relation_field?: RecordIdString[]
	post_relation_field?: RecordIdString
	select_field_no_values?: string
	rich_editor_field?: HTMLString
}

export type MyViewRecord<Tjson_field = unknown> = {
	post_relation_field?: RecordIdString
	text_field?: string
	json_field?: null | Tjson_field
}

export type PostsRecord = {
	field?: string
	nonempty_field: string
	nonempty_bool: boolean
	field1?: number
}

export type UsersRecord = {
	name?: string
	avatar?: string
}

// Response types include system fields and match responses from the PocketBase API
export type BaseResponse = BaseRecord & BaseSystemFields
export type CustomAuthResponse = CustomAuthRecord & AuthSystemFields
export type EverythingResponse<Tanother_json_field = unknown, Tjson_field = unknown, Texpand = unknown> = EverythingRecord<Tanother_json_field, Tjson_field> & BaseSystemFields<Texpand>
export type MyViewResponse<Tjson_field = unknown, Texpand = unknown> = MyViewRecord<Tjson_field> & ViewSystemFields<Texpand>
export type PostsResponse = PostsRecord & BaseSystemFields
export type UsersResponse = UsersRecord & AuthSystemFields

export type CollectionRecords = {
	base: BaseRecord
	custom_auth: CustomAuthRecord
	everything: EverythingRecord
	my_view: MyViewRecord
	posts: PostsRecord
	users: UsersRecord
}"
`;
